sql-plugin/src/main/java/com/nvidia/spark/rapids/ColumnarCopyHelper.java

```diff
@@ -0,0 +1,196 @@
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) b.append(cv.getByte(i));
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.append(cv.getByte(i));
+    }
+  }
+
+  public static void shortCopy(ColumnVector cv, ColumnBuilder b, int rows) {
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) b.append(cv.getShort(i));
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.append(cv.getShort(i));
+    }
+  }
+
+  public static void intCopy(ColumnVector cv, ColumnBuilder b, int rows) {
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) b.append(cv.getInt(i));
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.append(cv.getInt(i));
+    }
+  }
+
+  public static void longCopy(ColumnVector cv, ColumnBuilder b, int rows) {
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) b.append(cv.getLong(i));
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.append(cv.getLong(i));
+    }
+  }
+
+  public static void floatCopy(ColumnVector cv, ColumnBuilder b, int rows) {
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) b.append(cv.getFloat(i));
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.append(cv.getFloat(i));
+    }
+  }
+
+  public static void doubleCopy(ColumnVector cv, ColumnBuilder b, int rows) {
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) b.append(cv.getDouble(i));
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.append(cv.getDouble(i));
+    }
+  }
+
+  public static void stringCopy(ColumnVector cv, ColumnBuilder b, int rows) {
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) b.appendUTF8String(cv.getUTF8String(i).getBytes());
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.appendUTF8String(cv.getUTF8String(i).getBytes());
+    }
+  }
+
+  public static void decimal32Copy(ColumnVector cv, ColumnBuilder b, int rows,
+      int precision, int scale) {
+    if (!cv.hasNull()) {
+      for (int i = 0; i < rows; i++) {
+        b.append((int) cv.getDecimal(i, precision, scale).toUnscaledLong());
+      }
+      return;
+    }
+    for (int i = 0; i < rows; i++) {
+      if (cv.isNullAt(i)) {
+        b.appendNull();
+        continue;
+      }
+      b.append((int) cv.getDecimal(i, precision, scale).toUnscaledLong());
```

Comment:
Can we file a follow on issue for the decimal types, or possibly put in the improvements now? If we know that the `ColumnVector` is a `WritableColumnVector` we could bypass the need to create a `Decimal` object at all and copy the data/bytes directly.

For decimal32 we could call into `intCopy`. For decimal64 we could call into `longCopy`, and for 128-bit decimal we could get the raw binary bytes.

https://github.com/apache/spark/blob/c8b34ab7340265f1f2bec2afa694c10f174b222c/sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java#L393-L397

and the open up an API in CUDF JNI to set those raw bytes.  This is what CUDF is doing already

https://github.com/rapidsai/cudf/blob/7025c40c16bfea9b37ac9110f62b7b23a2c51725/java/src/main/java/ai/rapids/cudf/HostColumnVector.java#L1313-L1315

Which is why I said it might be good as a follow on issue.

Reply:
Done.